### ISSUE TITLE
Guild missing in cache for no apparent reason.

### ISSUE THREAD

**Issue by LuckyMika** [serenity-rs/serenity#2590]
TLDR;
When trying to get a members permissions from a text message they sent, serenity always fails with `Model(GuildNotFound)`.
## Clarifications
The `cache` feature is enabled.
The `GUILD_MEMBERS`, `MESSAGE_CONTENT`, `GUILD_MESSAGES` and `GUILDS` intents are enabled.
All of the intents are turned on in the developer page of the application
The bot is in exactly 1 server

## The issues
When receiving a message from a guild member, I have tried to get their permissions with the following methods:
- Using the `member` field to get it's `permissions` field directly.
- Querying the member via the `member` function 
    - getting its `permission` field
    - querying the permissions via the `permissions` function
- Try to get the member via the guild's `member` function
    - getting its `permission` field
    - querying the permissions via the `permissions` function

## The error
All of the above have failed with an error stating that the guild cannot be found in the cache.
It happens when trying to `unwrap()` the permissions

## Example code
Here is some example code that I have made to confirm all of my tests
The code actually **fails**, particularly when fetching the guild, as it was not found in the cache. This makes no sense for me, since the cache feature is enabled, the message was sent in a guild and I have been testing this with the bot running for around 20 minutes so it couldn't even be a "cold" cache.
```rust
use serenity::{client::Context, model::channel::Message};

pub async fn test(ctx: Context, msg: Message) {
    println!("{:?}", msg.guild_id.expect("Message was not sent in a guild"));

    println!("{:?}", msg.clone().member.unwrap().permissions);
    println!("{:?}", msg.clone().member(&ctx).await.unwrap().permissions);
    println!("{:?}", msg.member(&ctx).await.unwrap().permissions(&ctx));
    println!("{:?}", msg.guild(&ctx).expect("Failed to fetch guild").member(&ctx, &msg.author.id).await.expect("Failed to fetch member").permissions);
    println!("{:?}", msg.guild(&ctx).expect("Failed to fetch guild").member(&ctx, &msg.author.id).await.expect("Failed to fetch member").permissions(&ctx));

    msg.reply(&ctx, "Successfully ran tests").await.expect("Failed to send message");
}
```
Additionally, here is my Cargo.toml:
```toml
[package]
name = "rust_bot"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
serenity = { default-features = false, version = "0.11", features = ["builder", "cache", "collector", "client", "model", "time", "http", "utils", "rustls_backend"] }
tokio = { version = "1.21.2", features = ["full"] }
dotenv = "0.15.0"
serde = "1.0.192"
serde_json = "1.0.108"
```

**Comment by arqunis** [comment]
I am unable to reproduce this. When running your test function on my bot, I get:
```
GuildId(<...>)
None
None
Ok(CREATE_INSTANT_INVITE | KICK_MEMBERS | BAN_MEMBERS | ADMINISTRATOR | MANAGE_CHANNELS | MANAGE_GUILD | ADD_REACTIONS | VIEW_AUDIT_LOG | PRIORITY_SPEAKER | STREAM | VIEW_CHANNEL | SEND_MESSAGES | SEND_TTS_MESSAGES | MANAGE_MESSAGES | EMBED_LINKS | ATTACH_FILES | READ_MESSAGE_HISTORY | MENTION_EVERYONE | USE_EXTERNAL_EMOJIS | VIEW_GUILD_INSIGHTS | CONNECT | SPEAK | MUTE_MEMBERS | DEAFEN_MEMBERS | MOVE_MEMBERS | USE_VAD | CHANGE_NICKNAME | MANAGE_NICKNAMES | MANAGE_ROLES | MANAGE_WEBHOOKS | MANAGE_EMOJIS_AND_STICKERS | USE_SLASH_COMMANDS | REQUEST_TO_SPEAK | MANAGE_EVENTS | MANAGE_THREADS | CREATE_PUBLIC_THREADS | CREATE_PRIVATE_THREADS | USE_EXTERNAL_STICKERS | SEND_MESSAGES_IN_THREADS | USE_EMBEDDED_ACTIVITIES | MODERATE_MEMBERS)
None
Ok(CREATE_INSTANT_INVITE | KICK_MEMBERS | BAN_MEMBERS | ADMINISTRATOR | MANAGE_CHANNELS | MANAGE_GUILD | ADD_REACTIONS | VIEW_AUDIT_LOG | PRIORITY_SPEAKER | STREAM | VIEW_CHANNEL | SEND_MESSAGES | SEND_TTS_MESSAGES | MANAGE_MESSAGES | EMBED_LINKS | ATTACH_FILES | READ_MESSAGE_HISTORY | MENTION_EVERYONE | USE_EXTERNAL_EMOJIS | VIEW_GUILD_INSIGHTS | CONNECT | SPEAK | MUTE_MEMBERS | DEAFEN_MEMBERS | MOVE_MEMBERS | USE_VAD | CHANGE_NICKNAME | MANAGE_NICKNAMES | MANAGE_ROLES | MANAGE_WEBHOOKS | MANAGE_EMOJIS_AND_STICKERS | USE_SLASH_COMMANDS | REQUEST_TO_SPEAK | MANAGE_EVENTS | MANAGE_THREADS | CREATE_PUBLIC_THREADS | CREATE_PRIVATE_THREADS | USE_EXTERNAL_STICKERS | SEND_MESSAGES_IN_THREADS | USE_EMBEDDED_ACTIVITIES | MODERATE_MEMBERS)
```
This is as soon as the bot booted up and I sent a message. And of course a `Successfully ran tests` message in the channel. I used your list of gateway intents.

Those `None`s are normal; the `Member::permissions` field is only populated if the message came from an interaction, that is, a slash command. In the `Ok` cases, the `Member::permissions()` method can successfully find the `Guild` object, and thus be capable of calculating the member's permissions.

Where are you calling `test()`? In the `EventHandler::message` handler? 
Are you able to access the guild from the cache yourself? `Member::permissions()` uses the [`Cache::guild_field()`](https://docs.rs/serenity/latest/serenity/cache/struct.Cache.html#method.guild_field) method.

**Comment by LuckyMika** [comment]
I'll update my code to include some comments for clarifications and upload the entire project when I'm home.

**Comment by LuckyMika** [comment]
I have uploaded my entire codebase to [this](https://github.com/LuckyMika/rusty-bot/) repo. There are a lot of features to be finished/polished, but right now I just want to get this issue resolved.
